core/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2TokenRevocationAuthenticationProvider.java

```diff
@@ -0,0 +1,107 @@
+import org.springframework.security.oauth2.server.authorization.OAuth2Authorization;
+import org.springframework.security.oauth2.server.authorization.OAuth2AuthorizationAttributeNames;
+import org.springframework.security.oauth2.server.authorization.TokenType;
+import org.springframework.security.oauth2.server.authorization.client.RegisteredClient;
+import org.springframework.security.oauth2.server.authorization.OAuth2AuthorizationService;
+import org.springframework.security.oauth2.server.authorization.client.RegisteredClientRepository;
+import org.springframework.util.Assert;
+
+/**
+ * @author Vivek Babu
+ */
+public class OAuth2TokenRevocationAuthenticationProvider implements AuthenticationProvider {
+
+	public static final String TOKEN_TYPE_HINT = "token_type_hint";
+	public static final String UNSUPPORTED_TOKEN_TYPE = "unsupported_token_type";
+
+	private RegisteredClientRepository registeredClientRepository;
+	private OAuth2AuthorizationService authorizationService;
+
+	/**
+	 * Constructs an {@code OAuth2TokenRevocationAuthenticationProvider} using the provided parameters.
+	 *
+	 * @param registeredClientRepository the repository of registered clients
+	 * @param authorizationService the authorization service
+	 */
+	public OAuth2TokenRevocationAuthenticationProvider(RegisteredClientRepository registeredClientRepository,
+			OAuth2AuthorizationService authorizationService) {
+		Assert.notNull(registeredClientRepository, "registeredClientRepository cannot be null");
+		Assert.notNull(authorizationService, "authorizationService cannot be null");
+		this.registeredClientRepository = registeredClientRepository;
+		this.authorizationService = authorizationService;
+	}
+
+	@Override
+	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+		OAuth2TokenRevocationAuthenticationToken authenticationToken = (OAuth2TokenRevocationAuthenticationToken)
+				authentication;
+
+		String clientId = authenticationToken.getPrincipal().toString();
+		RegisteredClient registeredClient = this.registeredClientRepository.findByClientId(clientId);
+
+		if (registeredClient == null) {
+			throw new OAuth2AuthenticationException(new OAuth2Error(OAuth2ErrorCodes.INVALID_CLIENT));
+		}
+
+		final String tokenTypeHint = authenticationToken.getTokenTypeHint();
+		final String token = authenticationToken.getToken();
+		final OAuth2Authorization authorization = findAuthorizationUsingHint(tokenTypeHint, token);
+
+		OAuth2TokenRevocationAuthenticationToken successfulAuthentication =
+				new OAuth2TokenRevocationAuthenticationToken(token, registeredClient, tokenTypeHint);
+
+		if (authorization == null) {
+			return successfulAuthentication;
+		}
+
+		if (!registeredClient.getClientId().equals(authorization.getRegisteredClientId())) {
+			throw new OAuth2AuthenticationException(new OAuth2Error(OAuth2ErrorCodes.INVALID_GRANT));
+		}
+
+		if (token.equals(authorization.getAttribute(OAuth2AuthorizationAttributeNames.CODE))) {
```

Comment:
It's not clear to me what this check is for?

Reply:
This is not needed Joe. I have removed it. As of now we ignore the hint and only look for access token